libs/datamodel/core/tests/attributes/relations/referential_actions/cycle_detection.rs

```diff
@@ -0,0 +1,632 @@
+        [1;94m25 | [0m    writtenBy   User     @relation(fields: [writtenById], references: [id])
+        [1;94m26 | [0m    [1;91mpost        Post     @relation(fields: [postId], references: [id])[0m
+        [1;94m27 | [0m}
+        [1;94m   | [0m
+    "#]];
+
+    expect.assert_eq(&datamodel::parse_schema(dml).map(drop).unwrap_err());
+}
+
+#[test]
+fn multiple_cascading_complex() {
+    let dml = indoc! {
+        r#"
+        datasource test {
+            provider = "sqlserver"
+            url      = "sqlserver://localhost:1433;database=master;user=SA;password=<YourStrong@Passw0rd>;trustServerCertificate=true"
+        }
+
+        model Address {
+            id        Int       @id @default(autoincrement())
+            users     User[]
+        }
+
+        model User {
+            id        Int       @id @default(autoincrement())
+            addressId Int
+            address   Address   @relation(fields: [addressId], references: [id])
+            comments  Comment[]
+            posts     Post[]
+            cements   Cement[]
+        }
+
+        model Post {
+            id        Int       @id @default(autoincrement())
+            authorId  Int
+            author    User      @relation(fields: [authorId], references: [id])
+            comments  Comment[]
+            tags      Tag[]     @relation("TagToPost")
+            cements   Cement[]
+        }
+
+        model Cement {
+            id          Int       @id @default(autoincrement())
+            postId      Int
+            userId      Int
+            tagId       Int
+            post        Post      @relation(fields: [postId], references: [id])
+            user        User      @relation(fields: [userId], references: [id])
+            tag         Tag       @relation(fields: [tagId], references: [id])
+            comments    Comment[]
+        }
+
+        model Comment {
+            id          Int      @id @default(autoincrement())
+            writtenById Int
+            postId      Int
+            cementId    Int
+            writtenBy   User     @relation(fields: [writtenById], references: [id])
+            post        Post     @relation(fields: [postId], references: [id])
+            cement      Cement   @relation(fields: [cementId], references: [id])
+        }
+
+        model Tag {
+            id      Int    @id @default(autoincrement())
+            tag     String @unique
+            posts   Post[] @relation("TagToPost")
+            cements Cement[] 
+        }
+    "#};
+
+    let expect = expect![[r#"
+        [1;91merror[0m: [1mError parsing attribute "@relation": When any of the records in models `Address`, `User` are updated or deleted, the referential actions on the relations cascade to model `Cement` through multiple paths. Please break one of these paths by setting the `onUpdate` and `onDelete` to `NoAction`. (Implicit default `onUpdate`: `Cascade`) Read more at https://pris.ly/d/cyclic-referential-actions[0m
```

Comment:
I tried to follow the paths in the datamodel above and I can't see why this triggers an error: there is only one relation on a single path between `Address` and `User`.

Reply:
But Address cascades User, and User is the culprit. So deleting Address will trigger the actions in User. It is slightly off, that we should just tell about the User model, but it's extremely hard to get that Address out from the error...

Reply:
Ok, let's just document that as a limitation in the implementation comments then.

Reply:
I should write the docs soon...